queenbee/schema/function.py

```diff
@@ -0,0 +1,151 @@
+    )
+
+    env: Dict[str, str] = Schema(
+        None,
+        description='A dictionary of key:values for environmental variables.'
+    )
+
+    outputs: Arguments = Schema(
+        None,
+        description='List of output arguments.'
+    )
+
+    @validator('inputs')
+    def check_workflow_reference(cls, v):
+        input_params = v.parameters
+        ref_params = []
+        for param in input_params:
+            if not param.value:
+                continue
+            if 'workflow.' in param.value:
+                ref_params.append(param)
+        if len(ref_params) > 0:
+            params = ['{}: {}'.format(param.name, param.value) for param in ref_params]
+            warnings.warn(
+                'Referencing workflow parameters in a template function makes the'
+                ' function less reusable. Try using inputs / outputs of the function'
+                ' instead and assign workflow values in flow section when calling'
+                ' this function.\n\t- {}'.format('\n\t-'.join(params))
+            )
+        return v
+
+    @validator('command')
+    def check_command_referenced_values(cls, v, values):
+        # get references
+        match = var_parser(v)
+        if not match:
+            return v
+        # ensure referenced values are valid
+        func_name = values['name']
+        input_names = [param.name for param in values['inputs'].parameters]
+        # check inputs
+        cls.validate_variable(match, func_name, input_names)
+        return v
+
+    @validator('outputs')
+    def check_output_referenced_values(cls, v, values):
+        output_params = []
+
+        if v.parameters is not None:
+            output_params.extend([p for p in v.parameters])
+        if v.artifacts is not None:
+            output_params.extend([p for p in v.artifacts])
+
+        ref_params = []
+        for param in output_params:
+            if not param.path:
+                continue
+            if 'workflow.' in param.path:
+                ref_params.append(param)
+        if len(ref_params) > 0:
+            params = ['{}: {}'.format(param.name, param.path) for param in ref_params]
+            warnings.warn(
+                'Referencing workflow parameters in a template function makes the'
+                ' function less reusable. Try using inputs / outputs of the function'
+                ' instead and assign workflow values in flow section when calling'
+                ' this function.\n\t- {}'.format('\n\t-'.join(params))
+            )
+        # check the variables are fine
+        func_name = values['name']
+        input_names = [param.name for param in values['inputs'].parameters]
+        for v in output_params:
+            match = var_parser(v.path)
+            cls.validate_variable(match, func_name, input_names)
+
+        return v
+
+    @staticmethod
+    def validate_variable(variables, func_name, input_names):
+        """Validate referenced variables."""
```

Comment:
It might be good to list what validation actually entails in the docstring. Also it seems odd to have a validation function that is not a `@validator`. Might be a good idea to explain why this is the case in the docstring.

Reply:
This is a helper method that is used by a number for `@validator` classmethods. That's why it is not created as a typical Pydantic validator. I added a better description to docstring.